### PR TITLE
[KYUUBI #1594] SparkProcessBuild submit add --conf spark.driver.host=

### PR DESCRIPTION
**Pull request by zwangsheng**
### _Why are the changes needed?_

detail #1594 
When Kyuubi On Kubernetes submit spark engine by client mode, executor pod can not connect to driver by kyuubi pod name, which is default set.

### _How was this patch tested?_
- [x] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [x] [Run test](https://kyuubi.readthedocs.io/en/latest/develop_tools/testing.html#running-tests) locally before make a pull request

### CODE REVIEW

**Comment by zwangsheng** [comment]
cc @yaooqinn @ulysses-you @pan3793 , please help check.

**Comment by yaooqinn** [comment]
why not just configure `spark.driver.host` directly?

**Comment by zwangsheng** [comment]
> why not just configure `spark.driver.host` directly?

When using kyuubi HA, we do not know driver will up on which node. So we can not point `spark.driver.host`  directly.

**Comment by zwangsheng** [comment]
And, local test build success. please retest.

**Comment by yaooqinn** [comment]
> > why not just configure `spark.driver.host` directly?
> 
> When using kyuubi HA, we do not know driver will up on which node. So we can not point `spark.driver.host` directly.

why, I saw you just point to a local host here

**Comment by zwangsheng** [comment]
> > > why not just configure `spark.driver.host` directly?
> > 
> > 
> > When using kyuubi HA, we do not know driver will up on which node. So we can not point `spark.driver.host` directly.
> 
> why, I saw you just point to a local host here

Kyuubi On Kubernetes use client deploy-mode, Spark Sql Engine will run in Kyuubi Pod, so use `InetAddress.getLocalHost.getHostAddress` get Kyuubi Pod IP.

**Comment by SimonYang-CS** [comment]
@zwangsheng, if we deploy Kyuubi as a statefulset app, then Kyuubi /Spark driver host can be reach from Spark executor.

[discussions 1606](https://github.com/apache/incubator-kyuubi/discussions/1606)

kyuubi-statefulset-example-2.kyuubi-server-headless.namespace.svc.cluster.local

which one do you prefer ? pod ip address or statefulset  host name ?

**Comment by zwangsheng** [comment]
Thanks for your idea, @SimonYang-CS ! I'll give it a try using `statefulset` kyuubi.

File: kyuubi-server/src/main/scala/org/apache/kyuubi/engine/spark/SparkProcessBuilder.scala [hunk 1]
```diff
@@ -157,6 +158,10 @@ class SparkProcessBuilder(
       buffer += s"$newKey=$v"
     }
 
+    buffer += CONF
```
Comment (yaooqinn) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/engine/spark/SparkProcessBuilder.scala:161]: conf.get(FRONTEND_THRIFT_BINARY_BIND_HOST).foreach

File: kyuubi-server/src/main/scala/org/apache/kyuubi/engine/spark/SparkProcessBuilder.scala [hunk 2]
```diff
@@ -157,6 +158,10 @@ class SparkProcessBuilder(
...
+    buffer += s"spark.driver.host=${conf.get(FRONTEND_THRIFT_BINARY_BIND_HOST)
+      .getOrElse(findLocalInetAddress.getCanonicalHostName)}"
```
Comment (yaooqinn) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/engine/spark/SparkProcessBuilder.scala:163]: getHostName?

**Comment by zwangsheng** [comment]
In test, `StatefulSet Kyuubi` can solve driver url problem. But we also need help Kyuubi in `Pod` or `Deployment` Run Spark Client Mode Success.

File: kyuubi-server/src/test/scala/org/apache/kyuubi/engine/spark/SparkProcessBuilderSuite.scala [hunk 3]
```diff
@@ -271,6 +271,26 @@ class SparkProcessBuilderSuite extends KerberizedTestHelper {
     assert(b1.toString.contains(s"--conf spark.files=$testKeytab"))
 
   }
+
+  test("Engine Bind on hostName or IP with Kyuubi") {
```
Comment (yaooqinn) [inline kyuubi-server/src/test/scala/org/apache/kyuubi/engine/spark/SparkProcessBuilderSuite.scala]: bind        
 host name

File: kyuubi-server/src/main/scala/org/apache/kyuubi/engine/spark/SparkProcessBuilder.scala [hunk 4]
```diff
@@ -157,6 +157,18 @@ class SparkProcessBuilder(
       buffer += s"$newKey=$v"
     }
 
+    /**
+     * Kyuubi respect user setting config, if user set `spark.driver.host`, will pass it on.
+     * If user don't set this, will use thrift binary bind host to set.
+     * Kyuubi wants the Engine to bind hostName or IP with Kyuubi.
```
Comment (yaooqinn) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/engine/spark/SparkProcessBuilder.scala:163]: Let's also explain the spark behavior of handling this conf

Comment (zwangsheng) [reply kyuubi-server/src/main/scala/org/apache/kyuubi/engine/spark/SparkProcessBuilder.scala:163]: OK, will add.

**Comment by yaooqinn** [comment]
thanks, merged to master and v1.4.1

**Comment by turboFei** [comment]
@zwangsheng 
if the deploy-mode is cluster, does it work?

**Comment by turboFei** [comment]
I see, it does not impact, please ignore it.

```
    // Ignore invalid spark.driver.host in cluster modes.
    if (deployMode == CLUSTER) {
      sparkConf.remove(DRIVER_HOST_ADDRESS)
    }
```

**Comment by bydeath** [comment]
If the configuration item `spark.driver.host` only works when there is only one instance pod? Because there is  a exact kyuubi pod   which references the executors.  But here are two kyuubi server pods default. And I am using the kyuubi of version 1.6.0.
Also there is only `kyuubi-deployment.yaml`, no statefulset yaml file.

**Pull request closed without merging**